### ISSUE TITLE
[CSA] False negative: a later `nonnull` variadic argument is missed after preceding calls

### ISSUE THREAD

**Issue by rikki322** [llvm/llvm-project#213558]
Hi, I found a false negative in `core.NonNullParamChecker`.

For a variadic function declared with `__attribute__((nonnull))`, CSA reports null arguments in earlier calls but misses a later null variadic argument after a short sequence of calls to the same function.

## Program

```objc
void testVararg(int k, void *p) {
  extern void testVararg_check(int, ...) __attribute__((nonnull));
  void *n = 0;

  testVararg_check(0);
  testVararg_check(1, p);

  if (k & 1)
    testVararg_check(1, n);

  testVararg_check(2, p, p);

  if (k == 2)
    testVararg_check(2, n, p);

  if (k == 3)
    testVararg_check(2, p, n);
}
```

## Comparison with the equivalent reported case

CSA already reports the null value when it is the first variadic pointer argument:

```objc
testVararg_check(2, n, p);
// warning: Null pointer passed to 2nd parameter expecting 'nonnull'
```

It misses the equivalent null value in the next variadic pointer position:

```objc
testVararg_check(2, p, n);
// no warning
```

The same variable `n` is definitely null in both calls, and the unnumbered `nonnull` attribute applies to both pointer arguments. The preceding calls are required to reproduce the inconsistency; without that call sequence, CSA reports the third parameter correctly. This makes the result dependent on earlier calls rather than only on the argument being checked.

## Actual result

Re-running CSA on the reproducer reports the earlier variadic null arguments:

```text
core.NonNullParamChecker line 9: Null pointer passed to 2nd parameter expecting 'nonnull'
core.NonNullParamChecker line 14: Null pointer passed to 2nd parameter expecting 'nonnull'
```

No `core.NonNullParamChecker` warning is emitted for line 17:

```objc
testVararg_check(2, p, n);
```

## Expected result

CSA should report a nonnull violation for the later variadic null argument regardless of the preceding calls.

## Verification

```sh
clang --analyze -x objective-c \
  -Xclang -analyzer-checker=core.NonNullParamChecker \
  -Xclang -analyzer-output=sarif \
  -o nonnull_variadic_third_argument_fn.sarif \
  nonnull_variadic_third_argument_fn.m
```

Version:

```text
clang version 24.0.0git (https://github.com/llvm/llvm-project.git 1cb7e838cd47ecad4050948c0c907ecb1f466ac3)
```

**Comment by steakhal** [comment]
tldr; works as intended and the test code you have demonstrates one important point:
The bug modeled as a fatal error - just like a segfault on dereferencing a nullptr.
This means that the report that is shown for `k & 1` kills the execution path preventing the engine to reach the `k == 3` path.
Because of this, you never reach the branch of `k == 3`, thus never discover the bug. This is the intended behavior to avoid getting spammed by redundant or infeasible bug reports preceded by another fatal bug.

Here is a slightly rewritten variant of your tests that demonstrate that the order of arguments doesn't matter:
https://godbolt.org/z/bqMnjxKe4
```c++
extern int check(int, ...) __attribute__((nonnull));
int top(int rng, void* p) {
    void* n = nullptr;
    switch (rng) {
        case 10: return check(0); // no-warning: TN
        case 20: return check(1, p); // no-warning: TN
        case 30: return check(1, n); // warn: TP
        case 40: return check(2, p, p); // no-warning: TN
        case 50: return check(2, n, p); // warn: TP
        case 60: return check(2, p, n); // warn: TP
    }
    return 0;
}
```
